src/main/java/org/dependencytrack/tasks/MetricsUpdateTask.java

```diff
@@ -57,585 +81,883 @@ public class MetricsUpdateTask implements Subscriber {
-            }
+        final var counters = new Counters();
+
+        final List<Long> activeProjectIds;
+        try (final PersistenceManager pm = PersistenceManagerFactory.createPersistenceManager()) {
+            LOGGER.debug("Fetching active projects");
+            activeProjectIds = getActiveProjects(pm);
+            LOGGER.debug("Portfolio metrics update will include " + activeProjectIds.size() + " projects");
         }
 
-        LOGGER.debug("Project iteration complete. Iterating through all project metrics");
-        // Iterate through the metrics from all project
-        for (final MetricCounters projectMetrics: projectCountersList) {
-            // Add individual project metrics to the overall portfolio metrics
-            portfolioCounters.projects++;
-            portfolioCounters.critical += projectMetrics.critical;
-            portfolioCounters.high += projectMetrics.high;
-            portfolioCounters.medium += projectMetrics.medium;
-            portfolioCounters.low += projectMetrics.low;
-            portfolioCounters.unassigned += projectMetrics.unassigned;
+        if (activeProjectIds.size() > 0) {
+            // In order to speed up the process of updating metrics for all projects,
+            // we parallelize the work using a fixed-size thread pool.
+            // The pool size is intentionally left conservative for now, as we want to
+            // limit the impact on other functions of DT. We may increase this in the future,
+            // or allow it to be configured by users.
+
+            int threadPoolSize = SystemUtil.getCpuCores();
+            if (threadPoolSize > activeProjectIds.size()) {
+                threadPoolSize = activeProjectIds.size();
+            }
+            LOGGER.debug("Using thread pool size of " + threadPoolSize);
 
-            // All vulnerabilities
-            portfolioCounters.vulnerabilities += projectMetrics.severitySum();
+            final ExecutorService executorService = Executors.newFixedThreadPool(threadPoolSize);
+            final var limitingExecutor = new LimitingExecutor(executorService, threadPoolSize);
+            final var projectCounterFutures = new HashMap<Long, CompletableFuture<Counters>>();
+            try {
+                for (final long projectId : activeProjectIds) {
+                    LOGGER.debug("Submitting metrics update task for project ID " + projectId);
+                    projectCounterFutures.put(projectId, CompletableFuture.supplyAsync(() -> {
+                        try {
+                            return updateProjectMetrics(projectId);
+                        } catch (NoSuchElementException e) {
+                            LOGGER.warn("Couldn't update project metrics because the project was not found. " +
+                                    "This typically happens when the project was deleted after the metrics update task started.", e);
```

Comment:
I added explicit handling for this case because it's fairly common, and to some extent expected. The previous implementation would log this as "unexpected error", causing confusion for some users.

It's now logged as:

```
2022-06-22 19:32:57,764 WARN [MetricsUpdateTask] Couldn't update project metrics because the project was not found. This typically happens when the project was deleted after the metrics update task started.
java.util.NoSuchElementException: Project with ID 2640 does not exist
	at org.dependencytrack.tasks.MetricsUpdateTask.lambda$updateProjectMetrics$1(MetricsUpdateTask.java:328)
	at java.base/java.util.Optional.orElseThrow(Optional.java:408)
	at org.dependencytrack.tasks.MetricsUpdateTask.updateProjectMetrics(MetricsUpdateTask.java:328)
	at org.dependencytrack.tasks.MetricsUpdateTask.lambda$updatePortfolioMetrics$0(MetricsUpdateTask.java:145)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at org.dependencytrack.common.LimitingExecutor.lambda$execute$0(LimitingExecutor.java:69)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

But considering that it's common, expected, and users can't do anything about it, maybe we should log this with `debug` level instead?